src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1653,6 +1652,9 @@ private class ConnectionState : IInvocationBinder
             public Exception CloseException { get; set; }
             public CancellationToken UploadStreamToken { get; set; }
 
+            // We store this task so we can view it in a dump file, but never await it
+            public Task InvocationMessageReceiveTask { get; set; }
```

Comment:
Should we await this before allowing the HubConnection to be restarted? Is there any risk in letting `.On` handlers for the previous connection to still execute after the connection restarted?

Reply:
Like reconnnects? Yes.. That makes sense

Reply:
Lets say they do something like:
```c#
hubConnection.On("Reset", async () =>
{
    await hubConnection.StopAsync();
    await hubConnection.StartAsync();
});
```

With the current version of the PR it will work, however if we wait for the handlers to complete when start is called this would hang still.

I'm still leaning towards not waiting for this task anywhere. There shouldn't be any issues in the library code with previous `.On` handlers running. There could be customers who expect a certain order of operations or something and they could be surprised I guess.

Reply:
I don't have a great fix for stopping then starting in an On handler. I don't think we should break that.

The cleanest solution I can think of is to probably continue awaiting invocationMessageReceiveTask at the end of the ReceiveLoop, but then timing it out. If it times out, we could throw an exception indicating a possible deadlock that suggests dispatching the call to StopAsync() if it's made in an On handler.

That's a pretty big change though, and this is a definite improvement.